src/terminal/init.lua

```diff
@@ -28,17 +29,41 @@ end
 
 
 local sys = require "system"
+-- load submodules
+-- local input = require "terminal.input"
+-- local output = require "terminal.output"
+-- local clear = require "terminal.clear"
+-- local scroll = require "terminal.scroll"
+
+-- M.input = input
+-- M.output = output
+-- M.clear = clear
+-- M.scroll = scroll
+
+-- Lazy loading submodules to avoid circular dependencies
+setmetatable(M, {
+  __index = function(self, key)
+    if key == "input" then
+      self.input = require("terminal.input")
+      return self.input
+    elseif key == "output" then
+      self.output = require("terminal.output")
+      return self.output
+    elseif key == "clear" then
+      self.clear = require("terminal.clear")
+      return self.clear
+    elseif key == "scroll" then
+      self.scroll = require("terminal.scroll")
+      return self.scroll
+    end
+  end
+})
```

Comment:
this is really nice! we can actually improve on it by parameterizing the modules:
```suggestion
do
  -- list of lazy-loadable modules
  local lazy_mods = {
  	input = true,
  	output = true,
  	clear = true,
  	scroll = true,
	}
	
	setmetatable(M, {
	  __index = function(self, key)
	    if not lazy_mods[key] then
	      return
	    end
	    
	    self[key] = require("terminal." .. key)
	    return self[key]
	  end
	})
end
```

Reply:
Glad to hear that. Parameterizing the modules sounds like a great idea, I'll definitely consider refining it further.